src/Internals/measures/structure.jl

```diff
@@ -261,12 +249,15 @@ true
 See also [`top_predators`](@ref) and [`trophic_classes`](@ref).
 """
 function trophic_levels(A::AbstractMatrix)
-    A = Bool.(A)
-    trophiclvl_dict = trophic_level(UnipartiteNetwork(A))
-    trophiclvl_val = trophiclvl_dict |> values |> collect
-    trophiclvl_keys = trophiclvl_dict |> keys
-    species_idx = parse.(Int64, [split(t, "s")[2] for t in trophiclvl_keys])
-    trophiclvl_val[sortperm(species_idx)]
+    A = Matrix(A) # Ensure A is dense for inversion.
+    S = size(A, 1) # Species richness.
+    out_degree = sum(A; dims = 2)
+    D = -(A ./ out_degree) # Diet matrix.
+    D[isnan.(D)] .= 0.0
+    D[diagind(D)] .= 1.0 .- D[diagind(D)]
+    # Solve with the inverse matrix.
+    inverse = iszero(det(D)) ? pinv : inv
+    inverse(D) * ones(S)
```

Comment:
It may be costly if you decide to build a great number of large food webs, but anyway the limiting factor remain the simulations of species dynamics I believe. Inversion cannot fail, it is guaranteed to succeed if `det(D) = 0`, and if not pseudo-inversion is used.